### DISCUSSION TITLE
imagickFormat SVG producess invalid SVG

### DISCUSSION THREAD

**Discussion by muuvmuuv** [chillerlan/php-qrcode#168] [General]
**Describe the bug**

I am trying to process the Imagick image after generation but with no possible option it seems right for me.

I have the imagick ext installed on PHP 8.1 and "libmagickcore-dev" (seems the one with SVG support).

This is the cachefile output:

```
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
  "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
<svg width="105" height="105">
stroke:#000000000000;fill:#000000000000;  <rect x="0" y="0" width="5" height="5"/>
stroke:#000000000000;fill:#000000000000;  <rect x="5" y="0" width="5" height="5"/>
stroke:#000000000000;fill:#000000000000;  <rect x="10" y="0" width="5" height="5"/>
stroke:#000000000000;fill:#000000000000;  <rect x="15" y="0" width="5" height="5"/>
```

This is the output SVG from Imagick:

```
<?xml version="1.0" standalone="no"?>
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
  "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
<svg width="105" height="105">
<g style="fill:black;fill-opacity:1;stroke:none;stroke-width:1;stroke-opacity:1;fill-rule:nonzero;">
<g style="">
  <rect x="0" y="0" width="5" height="5"/>
</g>
<g style="">
  <rect x="5" y="0" width="5" height="5"/>
</g>
```

**Code sample**
```php
<?php

use chillerlan\QRCode\QRCode;
use chillerlan\QRCode\QROptions;

$qr = new QRCode(new QROptions([
	'outputType' => QRCode::OUTPUT_IMAGICK,
	// 'eccLevel' => QRCode::ECC_M,
	'imagickFormat' => 'svg',
	'imagickBG' => '#fff',
	'imageTransparent' => true,
	'imageTransparencyBG' => [100, 100, 100],
	'addQuietzone' => false,
	'markupLight' => '#fff',
	'markupDark' => '#000',
]));

$blob = $qr->render($content);

$img = new \Imagick();
$img->readImageBlob($blob);

$img->writeImage($file->path . '.post.svg');

// $img->setOption('png:compression-level', 6);
$img->resizeImage($size, $size, \Imagick::FILTER_LANCZOS, 1);
// $img->setImageFormat('png');

$img->writeImage($file->path . '.png');

```

**Expected behavior**

**Screenshots**

**Environment (please complete the following information):**
 - PHP version/OS: PHP 8.1, latest Debian
 - Library version: latest

**Additional context**

**Comment by codemasher** [comment]
Hmm why would you try to output SVG via ImageMagick when you could use the [SVG output](https://github.com/chillerlan/php-qrcode/blob/7fd54b78d8207fbd175bfd312472ec644d42d22e/src/Output/QRMarkupSVG.php) instead? Rendering/converting SVG via IM is generally not recommended - you could use the [InkScape command line interface](https://superuser.com/questions/381125/how-do-i-convert-an-svg-to-a-pdf-on-linux) for proper conversion. Also, with your current settings you could just use the PNG output instead.

**Reply by muuvmuuv** [reply]
Because I want it always to have a 200x200 size (or based on user input) and Scale isn't what I'am looking for. IM looked like a good fit to resize the output without reading it afterwards or post-process.

**Reply by codemasher** [reply]
You can scale the SVG output to a fixed size via [`QROptions::svgWidth` and `QROptions::svgHeight`](https://github.com/chillerlan/php-qrcode/blob/main/src/QROptionsTrait.php#L141-L153). Alternatively, you can scale it via CSS properties or just put it in a `<div>` with a fixed size.

**Reply by muuvmuuv** [reply]
Yes ofc but I need it as PNG afterwards. Thats why I used IM. And anyway I have tried using SVG output but wasnt able to process it further. Best would be a blob or stream (dont know if php has that) to handle it through some other steps.

Another queastion but can post it later on a diff thread: adding a logo with the options provided, is that possible? I just saw adding a space for it but not adding an actual svg/png logo.

**Reply by codemasher** [reply]
Uh there is [`QROptions::$returnResource`](https://github.com/chillerlan/php-qrcode/blob/7fd54b78d8207fbd175bfd312472ec644d42d22e/src/QROptionsTrait.php#L211-L225) and [here's a resize example too](https://github.com/chillerlan/php-qrcode/blob/main/examples/imageWithLogo.php#L32-L60).

**Reply by muuvmuuv** [reply]
Ok I am sorry totally missed these two...

...the extra class for logo. Will test!

**Comment by muuvmuuv** [comment]
Final solution:

```php
$options = new QROptions([
    'imageBase64' => false,
    'outputType' => QRCode::OUTPUT_IMAGICK,
    'eccLevel' => QRCode::ECC_M,
    'returnResource' => true,
    'imageTransparent' => true,
    'addQuietzone' => false,
    'markupLight' => 'rgba(0,0,0,0)',
    'markupDark' => '#000000',
]);
$qr = new QRCode($options);

/** @var \Imagick */
$img = $qr->render($content);

$img->setImageFormat('png');
$img->resizeImage($size, $size, \Imagick::FILTER_GAUSSIAN, 0.1);
$img->setOption('png:compression-level', 6);

// Laravel can only display saved files
return $img->writeImage($file->path);
```
